### DISCUSSION TITLE
Code Refactor Design Decisions - any thoughts?

### DISCUSSION THREAD

**Discussion by ABaumher** [ABaumher/galaxy-integration-steam#14] [Q&A]
I'm looking into the code refactor while i have free time (my job toggles between lots of work and absolutely no work). Any ideas we want to bounce around to make this nicer can go here.

**Comment by ABaumher** [comment]
First up: Protobufs
What i really want to do is remove `protobuf_client` and `protocol_client` and replace them with a `protobuf_helper` which converts our stored data into the raw request message and then the raw response message back to readable data. My original idea was to convert the terrible generated protobuf python into strongly typed classes and then use those, but i'd be generating new objects twice for every message (once from packet to protobuf format, then again to the typed class). For some of these, this will still be required, but i don't want to do this for every message. I could add the mypy module to protoc so it generated nice type hints, but i'm mostly leaning toward using [betterproto](https://pypi.org/project/betterproto/) instead. I know this is an extra module, so i'd like some thoughts.

**Reply by ABaumher** [reply]
@slo @wwMRd @don-de-marco @TheSentry 

Feel free to tag anyone i missed or just randomly comment

**Reply by don-de-marco** [reply]
You don't want to modify any `*_pb2.py` ever. There's too much automagic going on and whenever you re-generate you'll have to re-create your typed classes. Also, if you're gonna rework the protobuf stuff, you might want to as well have a look into the RPC thing then. However, I have the suspicion that RPC might not work for us since our request and Steam's response are completely asynchronous.

I had a look into betterproto and their interfaces look much more accessible than the original protobuf's ones. It's also not an additional module since you're removing the original protobuf package in favor of betterproto (I don't think they're mutually exclusive but it's probably still a good idea to not use both at the same time). I personally don't have any experience with betterproto so that's just a gut feeling on my end.

**Reply by ABaumher** [reply]
Clarification: I'd be creating new classes that exist alongside the pb2 classes. When i wanted to pass those around I'd convert the pb2 classes into their nicely defined ones. The issue is I'd be doubling our memory use, for no good reason. 

Better protobuf stable (1.2.5) loses the import statements in the proto definitions. So, things like `import steammessages_base` don't appear in the generated code. Afaik we only need those for type hints because the class definitions don't hold data until runtime and at runtime python doesn't actually typecheck anything.
The only reason this could be an issue is that betterprotobuf also implements the grpc service stuff as well, and makes it work with async so i don't have to do that manually. This would let us remove all the dumb protobuf_client handler callbacks. If somehow it borks the nested messages that'd be an issue. I'm looking at adding "from \<x\> import \<y\>" to the classes so mypy and idle or pycharm or vs will be OK with the type hints of that's something i need to do

The beta seems to fix this but since none of the proto files have a package (normally python ignores packages because namespaces don't exist) it's all dumped to `__init__.py`. which is fine, i guess, is just massive and annoying to read.

**Reply by don-de-marco** [reply]
I don't understand your second paragraph. Can you please rephrase it?

**Reply by ABaumher** [reply]
Updated. Lmk if I'm not being verbose enough. I'm learning it as i go so my understanding might not be perfect or well defined/written

**Reply by don-de-marco** [reply]
The import statements for `steammesages_base` are generated by the protobuf compiler because there's an `import 'steammessages_base.proto'` in the proto file. If that statement wasn't there, the import wouldn't be there either. I'm almost certain that protoc doesn't skip unneeded imports but instead just "copies" them over to the generated py files (however, protoc does complain about missing imports, though). Type checking is impossible with the original protobuf since the members of the message classes are only generated at runtime and do not exist in the python files.

I don't see why there would be an issue with gRPC because of broken type checks. As you correctly said, Python doesn't type check anywhere (unless you tell it otherwise) and the rpc interface will probably just raise an exception at runtime when you provide an instance of an unexpected class/type. However, using gRPC should indeed make the protobuf_client obsolete since this class is basically a very basic implementation of what gRPC probably does as well.

Packages in python are defined by the path a file has. That's why protobuf just ignores namespace statements if compiling to python. The `__init__.py` thing you're mentioning only has an influence on how you import the message classes (`from proto.messages import MyMessage` vs `from proto.messages.mymessage import MyMessage`) but should not change any semantics.

**Comment by ABaumher** [comment]
Update: I started the conversion to what i'd roughly like in the start conversion branch (https://github.com/ABaumher/galaxy-integration-steam/tree/start_the_conversion) if you want to see what i'm thinking of. It's not even remotely usable at the moment but i'm done for the day and don't see why i wouldn't make a commit. I may make the parser class something that accepts a websocket and simply recreate it whenever we need to reconnect on another websocket. Figuring out how to generic on python was... fun, but should be really useful. 

At the moment i'm leaning toward making parser a websocket wrapper so i don't need to pass it in everywhere. our model sends the data to the parser, which fires it off. Then i have a "true" parser class that takes the results and converts them to useful data. Or this parsing could be done in the model, idk. 

I'm also leaning toward bubbling errors up through the futures so the code can actually be error checked. Futures automatically throw when they are cancelled, and since i have a list of them, it's possible when we shut down our socket to cancel any futures we're waiting on. For other errors, wrapping our parser calls in try catch and setting the future exception accordingly means we can pass that along to the model and actually get our errors. For errors we can recover from (like a shut down) the model code can connect on a new CM and replay the last message, which fixes the issue engiefox found where messages might get lost if we get a logoff `TryAnotherCM` while waiting for another message (that task will hang indefinitely, when doing login this is especially destructive). We only need to bubble up from the current parser class to our model. If need be, we can return a message to the controller saying "hey, we lost connection" so that we can ping GOG and not get a timeout, and then replay the last action there instead.

**Comment by ABaumher** [comment]
I think i'm going to necro this and use it for design flaws we find in the old code and how to address them.
@urwrstkn8mare welcome to the ... party?

if you or @don-de-marco can double check this for me that would be great - Can either of you find any use for https://github.com/ABaumher/galaxy-integration-steam/blob/master/src/leveldb_parser.py ? i did a search throughout my environment and couldn't find a reference. It says it was updated as part of 1.0.0 from FoG but i don't know _why_

**Reply by ABaumher** [reply]
part of public profiles. well, i guess i can remove it from 1.0.6

**Reply by don-de-marco** [reply]
yeah, looks to me like it was indeed only used in the public profile backend

**Comment by ABaumher** [comment]
Migrating from 1.1 Alpha:
https://github.com/ABaumher/galaxy-integration-steam/blob/master/src/steam_network/protocol_client.py#L372-L383

The log only checks the first 12K. What i think happen is that Steam sends us a response in the format of `Multi`. `Multi` is essentially a bunch of responses zipped together. so, when we go to parse this gigantic mess of PICS product, when it's too large to fit in one message they send a multi and our code parses it as two separate messages. The first message only contains the first 12, and any other one(s) contain the next 12k and so on. the problem is, the code expects all of them at once. if you have a massive library, your cache is invalidated every time.

**Comment by ABaumher** [comment]
I'm drawing a blank on where, but there are some set compares using integer ids that are converted to strings for reasons i don't understand. Most should be possible to use a lookup table (Dict[int, <whatever>]) and massively increase performance. If either of you want to point these out or work on cleanup, that'd be great

**Comment by ABaumher** [comment]
in games cache, since i'm cleaning that up to fix the beta issues - can anyone see where get_dlcs is used? i can't find anywhere it's used, and it should be part of the Game object we give GOG. now i understand why everyone else who looked at the code gave up.

**Reply by don-de-marco** [reply]
Seems to not be used but I agree that we should be returning DLCs as part the game list because Galaxy expects it

**Comment by ABaumher** [comment]
Achievements Cache and by extension, Cache, both seem to be unused. is there something i'm missing there? because we still get achievements afaik but the code doesn't use the top-level functions there. is this a FoG remnant from public profiles or something? it's in the steam_network code

**Reply by ABaumher** [reply]
afaik we don't cache achieves, which makes some sense because we can't compare them and we can't batch them, the only way to get achievements is to run a query at which point the old cached values become irrelevant. most of the plugin functionality is tied to you being online (you *may* stil be able to play games), so it doesn't make a ton of sense to cache them if you won't be able to see them anyway when offline. 

Personally, if it's possible to *view* that info while offline i'd think it'd be a good idea to cache it but idk if that's even available. If i can get away with not having the "Cache" class and its fingerprint nonsense, i'd very much prefer to.

**Reply by wwMRd** [reply]
Achievements are pulled with game stats and then retrieved directly from there

**Reply by ABaumher** [reply]
And from there it's stored into the game stats cache, ty. So, we don't need achievements cache, not do we need cache because even if they did anything they'd be redundant. Removed on my development repo

**Comment by don-de-marco** [comment]
@ABaumher If you have a minute, I'd like to hear your opinion on my attempt to introduce some potential dumbfuckery: https://github.com/ABaumher/galaxy-integration-steam/pull/27/commits/7361acfcd53ac182d6c9f7aa5077c115c79b02b3

**Reply by ABaumher** [reply]
Responded on the commit.

**Reply by don-de-marco** [reply]
Rewrite complete -> https://github.com/ABaumher/galaxy-integration-steam/pull/28

**Comment by urwrstkn8mare** [comment]
if it helps im working on cleaning up the codespace without any major logic changes so it might be easier to do reworks but lmk if i shouldn't

**Reply by don-de-marco** [reply]
@ABaumher already started doing some work, maybe you can help out there: https://github.com/abaumher/galaxy-integration-steam/tree/start_the_conversion

**Reply by ABaumher** [reply]
I'm more working on shifting the whole code structure to be less terrible. For the most part that means stripping out all the code that actually does anything, and after i finalize h how it'll all interact, actually implement the code. If you're able to clean up the code so that when i do go to do that, the code has already been rewritten or cleaned up that'll be massively helpful. 

Tl;Dr keep doing what you're doing, lol

**Reply by urwrstkn8mare** [reply]
yeah i thought `start_the_conversion` was supposed to be sort of your playground (hence `start_the`). i've been working off the `better_proto` branch but should i be working off `start_the_conversion`?

**Reply by ABaumher** [reply]
I'm hoping i can get the overall structure of the code done, document it, then let you both take a look and suggest improvements. Then, after that, I'll go back in and reimplement the actual logic without the bad design decisions (i.e. use a lookup table instead of iterating over a list, etc)

But if that last part is done for me i will not complain, lol

**Reply by urwrstkn8mare** [reply]
one thing i noticed `start_the_conversion` and `better_proto` are divergent. what are you planning on doing about merging the logic differences between the branches?

oh yeah i haven't finished. i basically started off by stripping out junk and consolidating code that had be repeated/spread then simultaneously reorganising the logic into better classes. 

don't take this as any sort of pressure but how long do you think it would take? just cause if you're almost done i'll just wait but if not i'll try to help in small amounts through PRs to `start_the_conversion`

**Reply by ABaumher** [reply]
No, work off better proto. Conversion is super broken. The commit messages also make no sense because they're just end of day builds. I'm probably just going to squash them.

I think i can get the conversion part ~50% done today, enough to actually show it to you all and maybe start working off of? 
At the moment i have plans for the skeleton MVC all the way up and through login, and it's mostly ready to implement.  
The caching of any messages steam sends us is not prototyped, (i'm planning for one class that stores each individual cache and handles writing them to the db file. Basically it just consolidates that logic into one location instead of all over like it is now. And with nicer function names). But i haven't gotten there yet. 

An aside, the login flow is what i wrote so cleaning up the code there will hopefully be straightforward. That said, it's mostly copied from the original code design so it's still terrible, lol

**Reply by urwrstkn8mare** [reply]
> maybe start working off of?

so the plugin would be in the normal working state?

either way no rush just let us know when its in a presentable state and we'll have a look and work it out

**Reply by ABaumher** [reply]
If we want to keep the plugin working all throughout the conversion (good idea so we don't introduce bugs), that's an option while i keep mucking about in the start conversion branch. I have a protocol handler class in the conversion branch that is ready to go. You could port the existing calls to use it instead (it removes the need for storing all those furures and callbacks in protocol client and protobuf client) and it should still all work, just a bunch cleaner. 

The job list code world need to be properly converted. Let me bring that to better proto deps and let you all integrate it in our existing code.

**Reply by urwrstkn8mare** [reply]
yeah i don't wanna block you so perhaps just keep working on your work until you are able to merge `better_proto` into your conversion and have it somewhat work?

**Comment by ABaumher** [comment]
Here's a rough outline of my new structure: 
The plugin handles any system specific code (finding what games you have installed, launching them, etc). Any calls to steam are passed to the MVC, specifically the controller. Controller them goes to the model, which writes a message, sends it, and waits for the result. Sends it back to controller and that process repeats. For login, the controller may have to go to the view instead, then take the results to the model, like standard mvc
It works great for messages we expect, but once we login, steam sends us messages we don't ask for. In the new logic, I'm calling these "unsolicited". These are passed to the (not implemented) cache which stores them until the gog plugin needs them. Obviously, if we don't need a given unsolicited message we can just ignore it.

**Reply by ABaumher** [reply]
Am aside, the recv task run loop is literally 3 lines.  As a result, it should never throw an exception, which is a lot nicer than we have now.

Let me get actually awake and I'll get that all done

**Reply by urwrstkn8mare** [reply]
isn't the MVC model about apps with the user interface? we're just a plugin to gog galaxy not really a full-fledged app. 

Also i like the idea of the controller and if lets say that would be `client` module then we could also have a module for the system specific code you described instead of having it in `plugin.py` - this incl system reg stuff.
So if the view is basically all the auth logic i'm guessing that would be a module by itself? that uses `client`
The rest of the stuff like caching and stuff is in a `utils` but i'm not sure where that fits in
What is the benefit of having a seperate model thing is it just basically submodule that abstracts protobuf stuff?

btw im not familiar with MVC model so take my comments with a grain of salt. just trying to make sure i'm on the right track

**Reply by ABaumher** [reply]
This isn't a perfect use of MVC but I learned pretty quick nothing in programming is completely pure (except haskell. Which is why no one uses it). 

you mentioned not knowing mvc well so here's a crash courseon how it's being used. Apologies if I'm dumbing this down too much.

MVC is typically used in ui-heavy code, you are correct, but we can still use the same strategy with our code that's less ui heavy. Our view object is responsible for things that deal with interacting with the user (through the client). This is done by creating the NextStep objects that gog will display a browser page for (login, 2fa) and parsing the results when those pages close. In the current code, that's handled by `backend_steam_network.py` and `DisplayURIHelper` (in tools iirc).

The controller is the middleman between gog client and steam. The tl:dr is it's given a job to do and determines who it needs to ask in order to complete the job. This usually requires multiple steps, so it may be getting results from the view (i.e. a username) and passing it to the model (in this case, asking for the rsa key for that username) or vice versa. 

The model is our backend storage. In most cases this is something like a database but for us it's the steam server. 

MVC is a means of "separation of concerns" which is just a fancy way of saying breaking your code up so each part only does one specific set of related jobs. If done properly it makes the code easier to understand. A simple analogy: a car's combustion engine (or electric vehicle's battery) shouldn't know how to turn on the car's lights, that's not its job.

**Reply by ABaumher** [reply]
To further illustrate that analogy, turning in the lights *may* require the engine to do something. It might, for example, require power. This is where the controller comes in (in modern vehicles, a computer). It tells the lights to turn on, the lights say they need power, then the controller tells the engine "make power and send it to the lights". 

The example is admittedly dumb but serves its purpose

**Reply by ABaumher** [reply]
First, question. Do you mean modules in the python sense (a collection of classes and functions designated by a folder and `__init__.py`) or in the "modular design" (each part a class that can be added in our dummied without much issue) sense? Or do you mean classes/objects and I'm just being an ignorant American (lol)?

I don't think we need full python modules for each part of mvc. The mvc code should be relatively elegant, and therefore one file/ class should be enough for each of these. The most code- heavy part of our plugin is sending messages, parsing the results, and caching anything we'll use later. These may be broken into their own class (sending and receiving messages is that protocol handler class i mentioned earlier) 

As for modular design, each component here is tightly integrated, but could be mocked out during development. In true oop i could have abstract base classes for model and view and let you drop in a dummy for testing but that's way overkill.

**Reply by urwrstkn8mare** [reply]
> First, question. Do you mean modules in the python sense (a collection of classes and functions designated by a folder and __init__.py) or in the "modular design" (each part a class that can be added in our dummied without much issue) sense? Or do you mean classes/objects and I'm just being an ignorant American (lol)?

no lmao ur not being ignorant. what i was thinking of was each module would be class that would store all the related state and methods. for example with system specific code a single file is suffecient like you said. but other parts can get pretty heavy and dumping it all into one file may be unwise. so it would be a folder with `__init__.py` that only exposes the final class to be used by `plugin.py`

for now though just ignore me better we discuss when i can back myself with examples

**Comment by ABaumher** [comment]
ok, i've updated better_proto_deps with the socket_handler. it's not intended to be used this way (it's expecting more cleanup), but it should be able to drop in to the existing code as a replacement for protobuf_client with some modifications to protocol_client. ideally, this is a stepping point to cleaning up and eventually removing protocol_client entirely (it can be put in websocket_client), but that's for another time. These changes won't speed up the code but they make it a lot cleaner.

@don-de-marco @urwrstkn8mare if you can take a look at it, lmk if anything looks like it was improperly done. Further, if you can implement it in so the existing code uses it without breaking, that'd be great. It lets us test to see if this class is broken anywhere before the larger rewrite hides it amongst everything else that changed. To implement this, you will need to update protocol client as well. I've added migration notes, but they aren't great. lmk if you have any questions. 

NOTE: Job list is getting excised, there's a proper awaitable call for each of them (i think). use that instead. Errors currently proc on the run loop if you implement unsolicited as it is in the current code, but they'll still be handled properly so it's not a big deal. i forgot to implement `GetLastPlayedTimes(self)` but it should be easy to look at how every other one is done and copy it from protobuf_client and update it with the new call format. 

i'm writing how to convert the rsa call here since i forgot to do that in the migration. This replaces both `get_rsa_public_key` and `_rsa_handler`. It also removes the `rsa_handler = self._rsa_handler` from the init call.
```
    async def get_rsa_pkey_v2(self, username: str) -> Tuple[bool, SteamPublicKey]:
        proto_result = await self._socket_handler.GetPasswordRSAPublicKey(username)
        result : EResult = proto_result.eresult
        data : CAuthentication_GetPasswordRSAPublicKey_Response = proto_result.body

        logger.info ("GOT RSA KEY IN PROTOCOL_CLIENT")
        #If you provide a bad username, it still returns "OK" and gives you rsa key data. i have no idea why. it just does. so we have no way to determine bad login. 
        if (result == EResult.OK):
            key = SteamPublicKey(PublicKey(data.publickey_mod, data.publickey_exp), data.timestamp)
            return (True, key)
        #the only way we get here afaik is if steam is down or busy or something network related. 
        else: 
        #    self._auth_lost_handler = auth_lost_handler
            logger.warning(f"Received unknown error, code: {result}")
            #at this point, hopefully key would be null, so the bool part of the tuple would be redundant. but i can't seem to reach this state so idk. 
            return (False, None)
```

**Reply by ABaumher** [reply]
(obviously if you use this, you'd need to rename it. this example removed the auth_lost callback, which you'd need to add back. It really needs to die. it's confusing at best, useless at worst. and we can just use set exception on a task, making it redundant. )

**Reply by urwrstkn8mare** [reply]
im probably not gonna mess with that side of the codebase yet because you and @don-de-marco are far more equipped to handle it. i'll try seeing if i can clean up anything else in the meanwhile while i learn how the steam network parts work.

**Comment by ABaumher** [comment]
@don-de-marco betterproto just cut a new release (2.0.0 beta 6) a few days ago. I haven't looked at it yet but maybe worth trying? Iirc one of the betas started supporting default values so maybe that's be useful, idk.

**Reply by don-de-marco** [reply]
Nah, don't put yourself into that. Wait at least for the first point release and let others find the bugs. We have other things to focus on

**Reply by ABaumher** [reply]
well 1.2.5 has all fixed-width integer types as `float` for reasons no one really knows. One of the first PRs merged in (to 2.0.0b1) was fixing it to `int`. Not really an issue as long as we know the correct type is int, but linters may complain about us coercing floating point as int wherever we deal with fixed 64. job ids are the big one since i need them to make the request/response mapper work. 
Incidentally, it's not working, and hasn't been since the original version. Reverse-engineering SteamKit's code (that stuff is obfuscated as hell. I'm equal parts in awe and disgusted) hopefully has the solution i need. i'm testing it now

**Reply by ABaumher** [reply]
If the job id hack fails i can work without it, but it's a little more magic and that's prone to errors. the beta release might also fix it, idk. i'll check that afterward.

**Comment by ABaumher** [comment]
Ok, my foray into using Multis to group messages was less than fruitful. It appears that in addition to using protobuf, everything is wrapped in a Multi (aka a compressed message) to save on bytes transferred for faster transmission/parsing/etc. This also means that multis are never nested, hopefully. 

There is one major caveat though: If multiple messages are ready at the same time, they may be packed together. It appears that as part of Client Log On, several messages are sent along with the login response in one multi. This includes, crucially, the License List, which means if it is broken into two parts when over 12k licenses, it should still be contained in the same Multi. It's theoretically possible for other messages to do this in the Client Log On Response group. I think this is useful in this limited case, so it's worth the slight code complexity. i'll need to figure out the best way to handle this.

The biggest issue we've been facing is that Steam Client is largely callback-driven and GOG's API is async. It's like trying to fit a round peg in a square hole. I've tried to find a nice way of knowing when an entire block of data is complete, but Steam sends data as soon as it's available, not in nice chunks, and afaik does not ever signal being done. It works there because they are designed to take the data in ad-hoc, but GOG's plugin is structured and prefers the data sequentially. We've done a bunch of workarounds (marking data "dirty", knowing we're done when all entries have been marked, etc), but if something is lost for whatever reason, we don't know to stop waiting for it. 

Currently, there are 3 messages that may be broken up and i'm not entirely sure how to address them. Possible solutions are sub-lists.
1. PICS Product Info: Needed to know what games are part of a package. ad-hoc nature means we could be awaiting for all the games an unknown amount of time. 
    a) Request contains a `single_response` bool, setting it _may_ work. 
    b) Response contains a `response_pending` bool, it may contain relevant info to know if more responses are incoming.
2. LicenseList: Needed to know what packages we have. Could be sent in parts if user has too many licenses.
    a) Part of the Client Log On Multi. Likely means even if it is split in parts, we would know it is finished when the Client Log On Multi is fully unpacked.
3. PersonaState. Needed to get metadata on a user (profile picture, login status). Requested once, but also sent all periodically.
    a) May not need immediately? Friend Ids are sent with client log on, If the username and profile are cached, we can just use them until we get an update saying otherwise.
    b) UserPresence is expected to update all the time anyway. We may just be able to let it come in and update it as needed.

**Reply by urwrstkn8mare** [reply]
> Request contains a single_response bool, setting it may work

this seems like it'd work

i'm not expereinced with this side of the code but ur solutions seem pretty legit

**Reply by ABaumher** [reply]
PICS Product Info apparently accepts are returns a jobid. SteamKit uses it in their sample code for async jobs (a paradigm i've been copying for the rework). if that works i'll be very happy as that'll make easily the most annoyig part of the old code much easier to work with. I'll have to do more in the backend to magic it but that's fine i guess

EDIT: they use the `response_pending` flag to determine if it's done. Hoo boy, that's helpful. Downside is they actually don't fire off the event until processing all PICS Product Info with that job id, my future logic fires off the first one. Which is actually really annoying and i don't have a clean way to handle it. I'll think of something but i don't know how clean i can get it without being like "if it's a PICs Product Info, do this special hack. If not, go about your business".

**Comment by ABaumher** [comment]
There's a potential issue with the more user-friendly send and await the response paradigm over the classic callback-driven design: There's a loop that is responsible for processing all the messages steam sends us, and if you wait for the results of another request in that loop, it'll deadlock. You're waiting for the results of another request, but a potential response can't be parsed while waiting for it. I don't want to make that entire block synchronous because then you lose things like `run_in_executor` or `sleep` (for task switching) and it'd potentially throttle the GOG code and cause a shutdown, so i've left that code async.  

There is a relatively straightforward solution in that you can wrap any calls that await another response in a task and await the result there, but the problem with wrapping things in a task is it essentially swallows errors. 

The issue arises with PICS Product Info, which bizarrely enough supports the job-based awaitable trick. It required rewriting that entire process (which i had to do for client calls anyway), but the license and app versions both are drastically simplified. The problem is I can't do send and await in the process loop after handling the license lists, because then i'm blocking that task from executing. This will also proc for UserPresence (from friend list). 

At the moment, i've sidestepped the issue by not giving the process object and the cache object a reference to the send/receive object, but the fact of the matter is that those calls still need to be made and i need a better solution. 
@don-de-marco @urwrstkn8mare @wwMRd if you have any thoughts lmk

**Reply by ABaumher** [reply]
The easiest option is making the gog calls do the send request/await response call, which is on a separate task to begin with, and the errors are sent back to GOG to handle. However, this means we're being reactive instead of proactive - we're not sending these calls until GOG asks for them, which means the end user has to wait longer for the data to be retrieved. 

I'm leaning toward doing this, albeit with a twist: Create separate tasks for these from the GOG login, just before it returns with a successful authentication. We can then await these tasks in their respective calls. This means they (potentially) start as soon as possible, and we can check to see if the tasks errored out when awaiting them in the calls. this prevents the errors from being swallowed.

**Comment by ABaumher** [comment]
All objects in `galaxy.api.types` utilize `@dataclass` which means all have a `to_dict()` call. We can use that with `json.loads` instead of creating our own custom serializable classes. in addition to making our caches simpler, it makes giving this data to GOG trivial - no conversions necessary.

**Reply by ABaumher** [reply]
some things, like package info, will still need our own classes since no analog for it exists in GOG. 
Additionally, they use string for ids instead of integer. I still think we could use them for storage, but our caches use lookup tables: `Dict[int, <whatever>` to make up for it. Their string counterpart is only used during de/serialization. We should be using lookup tables anyway to make finding and updating a particular game/friend constant time, but we don't need that efficiency when de/serializing the data, so we can just pull the Values from the dict and serialize them. `dump` and `load` already require getting _every_ piece of data so the overhead of generating a dict from a list or vice versa isn't going to kill us.

**Comment by ABaumher** [comment]
@urwrstkn8mare Question on your local code cleanup: MyPy checks mac.py on Windows and presumably windows.py on Mac - do you have a suggestion on how to get around this?

**Reply by ABaumher** [reply]
an aside, mypy support is going to be _unpleasant_ but i really want it so i stop going insane.

**Reply by urwrstkn8mare** [reply]
hmm have u tested it? cause i think it should only import mac.py or win.py into __init__.py depending on the platform but i'm not sure if mypy checks both anyway.

**Reply by ABaumher** [reply]
It errors out because client is expected to be the windows client type but the else sets it to Mac client type. I think it's due to mypy not knowing how to deal with `as` properly (it also doesn't like the as manifest iirc) but i'm not 100% sure. Ngl I've just ignored all the errors mypy spits out for anything in `local`

**Reply by urwrstkn8mare** [reply]
hmmm i'm alot busier now but i'll try to test it out asap. 

maybe its this? https://github.com/python/mypy/issues/8547#issuecomment-600931811 
would you be able to try replacing `IS_WIN = system() == "Windows"` in `local/shared.py` to `IS_WIN = sys.platform == 'win32'`. If that doesn't work I think we would have to replace all usages of IS_WIN with `sys.platform == 'wind32'`.
